モデルテスト、システムテスト追加

keiz1213 opened the pull request:
1. テストデータ作成をFactoryBotを使用
2. UserとReportのモデルテスト追加
3. 日報投稿機能とコメント投稿機能のシステムテストを追加

keiz1213 committed:
モデルテスト、システムテスト追加

1. テストデータ作成をFactoryBotを使用
2. UserとReportのモデルテスト追加
3. 日報投稿機能とコメント投稿機能のシステムテストを追加

cafedomancer commented on config/application.rb:
```diff
@@ -20,5 +20,14 @@ class Application < Rails::Application
+      g.factory_bot false
+    
+      # ファクトリファイルの置き場を変更
+      g.factory_bot dir: 'custom/dir/for/factories'
```
この設定だと `bin/rails factory_bot:model ...` を使用したときに `test/factories` ではないディレクトリにファクトリファイルが作成されてしまいそうな気がしました 👀

keiz1213 replied:
なるほどです。こちらの設定に関して、こちらの記事を参考にしました。https://qiita.com/morrr/items/f1d3ac46b029ccddd017#%E8%A8%AD%E5%AE%9A

この時特に何も調べることなくそのままコピペしてしまいました😅
改めて調べ直したところ、
```bash
$ bin/rails g factory_bot:model user
```
のようにジェネレータを使ってファクトリファイルを作成することがあるんですね💡
自分は手作業でファクトリファイル作ってました。。確かにこのような設定だとジェネレータを使用した場合変なとこに作成されそうですね。こちらの2つの設定を修正しておきます。ありがとうございます。

cafedomancer commented on test/factories/books.rb:
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :test_book, class: Book do
```
パフォーマンスの問題や読み込み順の問題が発生する可能性があるので、クラスの指定は文字列で行っておくのがオススメです！

> You can pass a constant as well, if the constant is available (note that this can cause test performance problems in large Rails applications, since referring to the constant will cause it to be eagerly loaded).

https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md#specifying-the-class-explicitly

keiz1213 replied:
公式ページもよく読んでなかったです😅
こちら改めて読みましたが、定数として渡すと色々問題が起きる可能性が有るのですね💡
勉強になります🙏

cafedomancer commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,55 @@
+
+  test 'destroying a report' do
+    visit reports_url
+    page.accept_confirm do
```
DSL として記載のあるメソッドは `page` を経由しなくても呼び出せると思うので試してみてください 💡 
```suggestion
    accept_confirm do
```

keiz1213 replied:
なるほどです。確かに`page`を省略しても呼び出せました💡

質問なのですが、

>DSL として記載のあるメソッド

とはどういう意味でしょうか？

ふわっとした質問で申し訳ないです🙏色々調べてみたのですが、どの資料も`page.accept_confirm`の形で記載されていて、`page`経由でない資料は見つけることができませんでした。

一応こちらでは
https://www.rubydoc.info/gems/capybara/Capybara/Session#accept_confirm-instance_method
`accept_confirm`を呼び出す際`page`が書かれていませんでしたが、このサイトでは逆になぜ`page`が書かれていないのか？という点もよくわかりません。よろしくおねがいします🙏

cafedomancer replied:
> 質問なのですが、
>
> > DSL として記載のあるメソッド
>
> ふわっとした質問で申し訳ないです🙏色々調べてみたのですが、どの資料も`page.accept_confirm`の形で記載されていて、`page`経由でない資料は見つけることができませんでした。

すみません、ドキュメントのリンクを貼り忘れていました...！ こちらが Capybara の DSL になります。 `accept_confirm` はここに載っていないのですが、`page` がなくても問題なく呼び出せます。理由はこのあとに書きます。
https://github.com/teamcapybara/capybara#the-dsl

> 一応こちらでは
https://www.rubydoc.info/gems/capybara/Capybara/Session#accept_confirm-instance_method
`accept_confirm`を呼び出す際`page`が書かれていませんでしたが、このサイトでは逆になぜpageが書かれていないのか？という点もよくわかりません。よろしくおねがいします🙏

Capybara には `Capybara::DSL` という mixin があって、これが include されているところでは上記の DSL をトップレベルで、つまり `page` なしで呼び出せるようになっています。上記の例は、この mixin が include されている想定で書かれているのかもしれません。

Rails のシステムテストで使うクラスは、元をたどると `ActionDispatch::SystemTestCase` というクラスを継承しており、このクラスの中で `Capybara::DSL` が include されているので、システムテストの中で `page` なしで DSL が使えるという仕組みになっています。

<img width="590" alt="Screen Shot 2022-02-18 at 20 54 12" src="https://user-images.githubusercontent.com/7645585/154678384-97410b55-2164-4827-980d-9601cfafa4d1.png">

https://api.rubyonrails.org/classes/ActionDispatch/SystemTestCase.html

DSL として使えるメソッドは、`Capybara::Session` クラスの `DSL_METHODS` という定数で定義されています。これに `accept_confirm` も含まれているので、`page` なしで `accept_confirm` を呼び出すことができるようになっています。

<img width="861" alt="Screen Shot 2022-02-18 at 20 58 12" src="https://user-images.githubusercontent.com/7645585/154678911-2c045dcc-7b69-4cc6-a3fc-173f95299ccb.png">

https://www.rubydoc.info/gems/capybara/Capybara/Session

cafedomancer commented on test/system/reports_test.rb:
```diff
@@ -0,0 +1,55 @@
+  test 'visiting the index' do
+    visit reports_url
+    assert_selector 'h1', text: '日報'
+  end
```
setup で first_report の日報を作っているので、その assertion があるとなお良いかな？と思いました。必須の修正ではないので余力があったら対応してみてください 💪

keiz1213 replied:
確かにです💡こちらassertionをつけました。ありがとうございます🙏確認お願いいたします。

keiz1213 committed:
指摘事項修正

1.RailsジェネレータのFactoryBotに関する設定削除
2.Factory定義時のクラス指定を定数から文字列に修正
3.レポート削除時のaccept_confirmメソッドをpageを省略して記述
4.レポートシステムテストのindex画面表示時にassertionを追加

cafedomancer reviewed (approved):
🌟
